File: mobile/src/utils/place-filters.ts
import type { TFunction } from 'i18next';

import i18n from '@/src/i18n';
import type { Place, PlaceCategory } from '@/src/data/places';
import { getPlaceOpenStatus } from '@/src/utils/place-hours';

type UserProfile = {
  profession?: string | null;
  interests?: string[];
  faith?: string | null;
};

function getProfileBoost(place: Place, profile: UserProfile): number {
  let score = 0;
  const tags = place.tags;
  const cat = place.category;
  const { profession, interests = [], faith } = profile;

  const isInterested = (interest: string) => interests.includes(interest);

  if (profession === 'architect' || isInterested('architecture')) {
    if (tags.includes('architecture') || cat === 'landmark' || cat === 'museum') score += 6;
  }
  if (profession === 'historian' || isInterested('history')) {
    if (cat === 'museum' || cat === 'cultural-spot' || cat === 'landmark' || tags.includes('history')) score += 6;
  }
  if (profession === 'photographer' || isInterested('photography')) {
    if (tags.includes('photogenic') || cat === 'viewpoint') score += 6;
  }
  if (profession === 'foodie' || isInterested('food')) {
    if (cat === 'cafe' || cat === 'restaurant' || tags.includes('meal') || tags.includes('coffee break')) score += 6;
  }
  if (profession === 'artist' || isInterested('art')) {
    if (tags.includes('art') || cat === 'museum' || cat === 'cultural-spot') score += 5;
  }
  if (isInterested('nature')) {
    if (cat === 'beach' || cat === 'nature' || cat === 'walking-area') score += 5;
  }
  if (isInterested('nightlife')) {
    if (tags.includes('nightlife') || tags.includes('date night')) score += 5;
  }
  if (isInterested('religion')) {
    if (tags.includes('religion') || cat === 'cultural-spot') score += 4;
  }
  if (faith && faith !== 'secular' && faith !== 'prefer_not_to_say') {
    if (tags.includes('religion') || cat === 'cultural-spot') score += 3;
  }

  return score;
}

export function sortPlacesForProfile(places: Place[], profile: UserProfile): Place[] {
  if (!profile.profession && !profile.interests?.length && !profile.faith) {
    return sortPlacesForBrowse(places);
  }
  return [...places].sort((a, b) => {
    const boostDiff = getProfileBoost(b, profile) - getProfileBoost(a, profile);
    if (boostDiff !== 0) return boostDiff;
    const openDiff = getOpenPriority(b) - getOpenPriority(a);
    if (openDiff !== 0) return openDiff;
    return getPlaceQualityScore(b) - getPlaceQualityScore(a);
  });
}

export type PlaceFilters = {
  query: string;
  category: PlaceCategory | 'all';
  tag: string | 'all';
  openNow: boolean;
};

const IMPORTANCE_SCORE = {
  hero: 3,
  supporting: 2,
  'long-tail': 1,
} as const;

export const CURATED_TAGS = [
  'rainy day',
  'short stop',
  'family',
  'budget',
  'local favorite',
  'waterfront',
  'photogenic',
  'date night',
  'coffee break',
  'meal',
] as const;

const TAG_LABEL_KEYS: Record<string, string> = {
  'rainy day': 'tags.rainyDay',
  'short stop': 'tags.shortStop',
  family: 'tags.family',
  budget: 'tags.budget',
  'local favorite': 'tags.localFavorite',
  waterfront: 'tags.waterfront',
  photogenic: 'tags.photogenic',
  'date night': 'tags.dateNight',
  'coffee break': 'tags.coffeeBreak',
  meal: 'tags.meal',
};

function defaultT(key: string, options?: Record<string, unknown>) {
  return i18n.t(key, options as any) as string;
}

export function formatTag(tag: string, t: TFunction = defaultT as TFunction): string {
  const key = TAG_LABEL_KEYS[tag];
  if (key) return t(key) as string;
  return tag.split(' ').map((w) => w[0].toUpperCase() + w.slice(1)).join(' ');
}

function compareTags(left: string, right: string) {
  const leftIndex = CURATED_TAGS.indexOf(left as (typeof CURATED_TAGS)[number]);
  const rightIndex = CURATED_TAGS.indexOf(right as (typeof CURATED_TAGS)[number]);
  const leftPriority = leftIndex === -1 ? Number.MAX_SAFE_INTEGER : leftIndex;
  const rightPriority = rightIndex === -1 ? Number.MAX_SAFE_INTEGER : rightIndex;

  if (leftPriority !== rightPriority) {
    return leftPriority - rightPriority;
  }

  return left.localeCompare(right);
}

export function getAllTags(places: Place[]) {
  const set = new Set<string>();
  for (const p of places) for (const t of p.tags) set.add(t);
  return Array.from(set).sort(compareTags);
}

export function getCuratedTags(places: Place[]) {
  const tags = new Set(getAllTags(places));
  return CURATED_TAGS.filter((tag) => tags.has(tag));
}

export function filterPlaces(places: Place[], filters: PlaceFilters) {
  const q = filters.query.trim().toLowerCase();
  return places.filter((p) => {
    if (filters.category !== 'all' && p.category !== filters.category) return false;
    if (filters.tag !== 'all' && !p.tags.includes(filters.tag)) return false;
    if (filters.openNow) {
      const status = getPlaceOpenStatus(p);
      if (status.state !== 'open' && status.state !== 'all-day') return false;
    }
    if (!q) return true;
    return (
      p.name.toLowerCase().includes(q) ||
      p.description.toLowerCase().includes(q) ||
      p.tags.some((t) => t.toLowerCase().includes(q))
    );
  });
}

function getOpenPriority(place: Place) {
  const status = getPlaceOpenStatus(place);

  switch (status.state) {
    case 'all-day':
      return 4;
    case 'open':
      return 3;
    case 'closed':
      return 2;
    case 'unknown':
      return 1;
    case 'temporarily-closed':
    default:
      return 0;
  }
}

function getVerifiedPriority(place: Place) {
  const status = getPlaceOpenStatus(place);
  let score = status.verified ? 2 : 0;
  if (place.image.verified) score += 1;
  return score;
}

export function getPlaceQualityScore(place: Place) {
  let score = 0;

  if (place.importanceTier === 'hero') score += 30;
  else if (place.importanceTier === 'supporting') score += 18;
  else score += 6;

  if (place.image.verified) score += 6;
  if (place.visitInfo?.hoursVerified) score += 4;

  if (typeof place.wiki?.confidence === 'number') {
    score += Math.min(14, Math.floor(place.wiki.confidence / 10));
  }

  if (place.wiki?.status === 'matched') score += 3;
  if (place.tags.includes('local favorite')) score += 2;
  if (place.tags.includes('photogenic')) score += 1;

  return score;
}

export function isHighQualityPlace(place: Place) {
  return getPlaceQualityScore(place) >= 24;
}

export function sortPlacesForBrowse(places: Place[]) {
  return [...places].sort((left, right) => {
    const openDiff = getOpenPriority(right) - getOpenPriority(left);
    if (openDiff !== 0) return openDiff;

    const verifiedDiff = getVerifiedPriority(right) - getVerifiedPriority(left);
    if (verifiedDiff !== 0) return verifiedDiff;

    const qualityDiff = getPlaceQualityScore(right) - getPlaceQualityScore(left);
    if (qualityDiff !== 0) return qualityDiff;

    const importanceDiff =
      IMPORTANCE_SCORE[right.importanceTier] - IMPORTANCE_SCORE[left.importanceTier];
    if (importanceDiff !== 0) return importanceDiff;

    return left.name.localeCompare(right.name);
  });
}

File: native/Piri/Sources/Utils/PlaceFilters.swift
import Foundation

/// Port of `mobile/src/utils/place-filters.ts`.
enum PlaceFilters {
    static let curatedTags = [
        "rainy day", "short stop", "family", "budget", "local favorite",
        "waterfront", "photogenic", "date night", "coffee break", "meal",
    ]

    private static let tagLabelKeys: [String: String] = [
        "rainy day": "tags.rainyDay",
        "short stop": "tags.shortStop",
        "family": "tags.family",
        "budget": "tags.budget",
        "local favorite": "tags.localFavorite",
        "waterfront": "tags.waterfront",
        "photogenic": "tags.photogenic",
        "date night": "tags.dateNight",
        "coffee break": "tags.coffeeBreak",
        "meal": "tags.meal",
    ]

    static func formatTag(_ tag: String) -> String {
        guard let key = tagLabelKeys[tag] else {
            return tag.split(separator: " ").map { $0.prefix(1).uppercased() + $0.dropFirst() }.joined(separator: " ")
        }
        return String(localized: String.LocalizationValue(key))
    }

    static func curatedTags(in places: [Place]) -> [String] {
        let present = Set(places.flatMap(\.tags))
        return curatedTags.filter { present.contains($0) }
    }

    private static func openPriority(_ place: Place) -> Int {
        switch PlaceHours.openStatus(for: place).state {
        case .allDay: return 4
        case .open: return 3
        case .closed: return 2
        case .unknown: return 1
        case .temporarilyClosed: return 0
        }
    }

    private static func verifiedPriority(_ place: Place) -> Int {
        var score = PlaceHours.openStatus(for: place).verified ? 2 : 0
        if place.image.verified { score += 1 }
        return score
    }

    static func qualityScore(_ place: Place) -> Int {
        var score = 0
        switch place.importanceTier {
        case .hero: score += 30
        case .supporting: score += 18
        case .longTail: score += 6
        }
        if place.image.verified { score += 6 }
        if place.visitInfo.hoursVerified { score += 4 }
        if let confidence = place.wiki?.confidence {
            score += min(14, Int(confidence / 10))
        }
        if place.wiki?.status == .matched { score += 3 }
        if place.tags.contains("local favorite") { score += 2 }
        if place.tags.contains("photogenic") { score += 1 }
        return score
    }

    static func isHighQuality(_ place: Place) -> Bool {
        qualityScore(place) >= 24
    }

    private static func importanceScore(_ tier: PlaceImportanceTier) -> Int {
        switch tier {
        case .hero: return 3
        case .supporting: return 2
        case .longTail: return 1
        }
    }

    static func sortedForBrowse(_ places: [Place]) -> [Place] {
        places.sorted { a, b in
            let openDiff = openPriority(b) - openPriority(a)
            if openDiff != 0 { return openDiff < 0 }
            let verifiedDiff = verifiedPriority(b) - verifiedPriority(a)
            if verifiedDiff != 0 { return verifiedDiff < 0 }
            let qualityDiff = qualityScore(b) - qualityScore(a)
            if qualityDiff != 0 { return qualityDiff < 0 }
            let importanceDiff = importanceScore(b.importanceTier) - importanceScore(a.importanceTier)
            if importanceDiff != 0 { return importanceDiff < 0 }
            return a.name.localizedCompare(b.name) == .orderedAscending
        }
    }

    private static func profileBoost(_ place: Place, profile: UserProfile) -> Int {
        var score = 0
        let tags = place.tags
        let category = place.category
        let interests = Set(profile.interests.map(\.rawValue))

        func interested(_ interest: String) -> Bool { interests.contains(interest) }

        if profile.profession == .architect || interested("architecture") {
            if tags.contains("architecture") || category == "landmark" || category == "museum" { score += 6 }
        }
        if profile.profession == .historian || interested("history") {
            if category == "museum" || category == "cultural-spot" || category == "landmark" || tags.contains("history") { score += 6 }
        }
        if profile.profession == .photographer || interested("photography") {
            if tags.contains("photogenic") || category == "viewpoint" { score += 6 }
        }
        if profile.profession == .foodie || interested("food") {
            if category == "cafe" || category == "restaurant" || tags.contains("meal") || tags.contains("coffee break") { score += 6 }
        }
        if profile.profession == .artist || interested("art") {
            if tags.contains("art") || category == "museum" || category == "cultural-spot" { score += 5 }
        }
        if interested("nature") {
            if category == "beach" || category == "nature" || category == "walking-area" { score += 5 }
        }
        if interested("nightlife") {
            if tags.contains("nightlife") || tags.contains("date night") { score += 5 }
        }
        if interested("religion") {
            if tags.contains("religion") || category == "cultural-spot" { score += 4 }
        }
        if let faith = profile.faith, faith != .secular, faith != .preferNotToSay {
            if tags.contains("religion") || category == "cultural-spot" { score += 3 }
        }
        if profile.budget == .budget, tags.contains("budget") { score += 5 }
        if profile.groupType == .family, tags.contains("family") { score += 6 }
        if profile.groupType == .couple, tags.contains("date night") { score += 5 }

        let duration = place.visitInfo.durationMinutes
        if profile.pace == .packed, tags.contains("short stop") || (duration.map { $0 <= 45 } ?? false) { score += 4 }
        if profile.pace == .relaxed, let duration, duration >= 90 { score += 4 }

        return score
    }

    private static func historyBoost(_ place: Place, viewed: [Place]) -> Int {
        var score = 0
        for other in viewed where other.id != place.id {
            if other.category == place.category { score += 2 }
            for tag in other.tags where place.tags.contains(tag) { score += 1 }
        }
        return min(score, 8)
    }

    static func sortedForProfile(_ places: [Place], profile: UserProfile, viewed: [Place] = []) -> [Place] {
        let hasProfile = profile.profession != nil || !profile.interests.isEmpty || profile.faith != nil
            || profile.budget != nil || profile.groupType != nil || profile.pace != nil
        guard hasProfile || !viewed.isEmpty else { return sortedForBrowse(places) }

        func totalBoost(_ place: Place) -> Int { profileBoost(place, profile: profile) + historyBoost(place, viewed: viewed) }

        return places.sorted { a, b in
            let boostDiff = totalBoost(b) - totalBoost(a)
            if boostDiff != 0 { return boostDiff < 0 }
            let openDiff = openPriority(b) - openPriority(a)
            if openDiff != 0 { return openDiff < 0 }
            return qualityScore(b) - qualityScore(a) < 0
        }
    }

    struct QueryFilters {
        var query: String = ""
        var category: String = "all"
        var tag: String = "all"
        var openNow: Bool = false
    }

    static func filter(_ places: [Place], with filters: QueryFilters) -> [Place] {
        let q = filters.query.trimmingCharacters(in: .whitespaces).lowercased()
        return places.filter { place in
            if filters.category != "all", place.category != filters.category { return false }
            if filters.tag != "all", !place.tags.contains(filters.tag) { return false }
            if filters.openNow, !PlaceHours.isOpen(PlaceHours.openStatus(for: place)) { return false }
            guard !q.isEmpty else { return true }
            return place.name.lowercased().contains(q)
                || place.description.lowercased().contains(q)
                || place.tags.contains { $0.lowercased().contains(q) }
        }
    }
}
